src/core/sync/event.d

```diff
@@ -0,0 +1,318 @@
+    import core.sys.windows.winbase /+: CreateEvent, CloseHandle, SetEvent, ResetEvent,
+        WaitForSingleObject, INFINITE, WAIT_OBJECT_0+/;
+}
+else version (Posix)
+{
+    import core.sys.posix.pthread;
+    import core.sys.posix.sys.types;
+}
+else
+{
+    static assert(false, "Platform not supported");
+}
+
+import core.time;
+import core.internal.abort : abort;
+
+/**
+ * This class represents an event. Clients of an event are suspended while waiting
+ * for the event to be "signaled".
+ *
+ * Implemented using `pthread_mutex` and `pthread_condition` on Posix and
+ * `CreateEvent` and `SetEvent` on Windows.
+ */
+struct Event
+{
+    // Posix version inspired by http://www.it.uu.se/katalog/larme597/win32eoposix
+nothrow @nogc:
+    /**
+     * Creates an event object.
+     *
+     * Params:
+     *  manualReset  = the state of the event is not reset automatically after resuming waiting clients
+     *  initialState = initial state of the signal
+     */
+    this(bool manualReset, bool initialState)
+    {
+        initialize(manualReset, initialState);
+    }
+
+    /**
+     * Initializes an event object.
+     *
+     * Params:
+     *  manualReset  = the state of the event is not reset automatically after resuming waiting clients
+     *  initialState = initial state of the signal
+     */
+    void initialize(bool manualReset, bool initialState)
+    {
+        version (Windows)
+        {
+            m_event = CreateEvent(null, manualReset, initialState, null);
+            m_event || abort("Error: CreateEvent failed.");
+        }
+        else version (Posix)
+        {
+            !pthread_mutex_init(cast(pthread_mutex_t*) &m_mutex, null) ||
+                abort("Error: pthread_mutex_init failed.");
+            m_state = initialState;
+            m_manualReset = manualReset;
+        }
+    }
+
+    @disable this(this);
+
+    ~this()
+    {
+        terminate();
+    }
+
+    void terminate()
+    {
+        version (Windows)
+        {
+            if (m_event)
+                CloseHandle(m_event);
+            m_event = null;
+        }
+        else version (Posix)
+        {
+            assert(!m_waiter);
+            pthread_mutex_destroy(&m_mutex) == 0 ||
+                abort("Error: pthread_mutex_destroy failed.");
+        }
+    }
+
+
+    /**
+     * Set the event to "signaled", so that waiting clients are resumed
+     */
+    void set()
+    {
+        version (Windows)
+        {
+            if (m_event)
+                SetEvent(m_event);
+        }
+        else version (Posix)
+        {
+            pthread_mutex_lock(&m_mutex);
+            m_state = true;
+            for (auto waiter = m_waiter; waiter != null; waiter = waiter.next)
+            {
+                pthread_mutex_lock(&waiter.mutex);
+                pthread_cond_signal(&waiter.cond);
+                pthread_mutex_unlock(&waiter.mutex);
+            }
+            pthread_mutex_unlock(&m_mutex);
```

Comment:
`scope(exit)`?

Reply:
The function is nothrow anyway, so it doesn't make a difference in  code generation. With a short function like this, having the symmetry show up in the function is slightly better IMO.